Introduce unreadCount store

markusstoll opened the pull request:
### ☑️ Implements issue https://github.com/nextcloud/spreed/issues/14667

* talk-desktop ref: https://github.com/nextcloud/talk-desktop/pull/1174

I had to introduce some changes to spreed that are described below to create a central unreadCountStore to have an even hook in talk-desktop to get updates of these unreadCounts and enable talk-desktop to set the correct badge count on the app icon.

### 1. New unreadCountStore
A new Vuex store module that serves as the single source of truth for all unread counters:

```javascript
state: {
    // Total counters
    totalUnreadMessages: 0,
    totalUnreadMentions: 0,
    totalUnreadMentionDirect: 0,

    // Per-conversation tracking
    unreadMessagesMap: {},
    unreadMentionsMap: {},
    unreadMentionDirectMap: {},
}
```

### 2. Integration Points

#### Conversations Store
- Enhanced `addConversation` mutation:
  - Automatically updates unread counters when adding new conversations
  - Handles initial state synchronization

- Improved `updateConversation` mutation:
  - Detects changes in unread counts
  - Dispatches updates to unreadCountStore
  - Maintains consistency across state changes

- Updated `deleteConversation` mutation:
  - Properly removes conversation counts from tracking
  - Updates total counters accordingly

#### Messages Store
- Enhanced `updateLastReadMessage` action:
  - Improved handling of "all messages read" scenarios
  - Proper counter reset when reaching conversation end
  - Better integration with chat-read-last feature

### 3. Event System
New event for badge updates:
```javascript
EVENTS.UNREAD_COUNT_UPDATED: 'talk:unread:updated'
```


### 🏁 Checklist

This issue is not browser depending.
But I added a matching change in talk-desktop and have a working application icon badge implementation.
This is already tested in Windows and MacOs builds of talk-desktop.

Pull request for talk-desktop is about to follow

## 🛠️ API Checklist

### 🚧 Tasks

- [ ] ...

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not possible
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 🔖 Capability is added or not needed

nickvergessen commented:
The main colleagues are currently out a bit. Hopefully they are able to give feedback towards the end of the week

markusstoll commented:
@Antreesy I understand your point, but unaware of your replacement solution for Vuex, I hardly could have taken that into account.
I pull request for talk-desktop is almost finished, you will have it shortly

Antreesy commented:
Don't worry about it, we appreciate efforts you put into this and will help with fixing that 😉
Thanks again for your contribution! I'll be back at Wednesday and will try to check it ASAP

Antreesy commented on src/store/unreadCountStore.js:
```diff
@@ -0,0 +1,182 @@
+		Vue.set(state.unreadMentionDirectMap, token, unreadMentionDirect || 0)
+
+		// Update totals
+		state.totalUnreadMessages += messageDiff
+		state.totalUnreadMentions += mentionsDiff
+		state.totalUnreadMentionDirect += mentionDirectDiff
```
getters should do this math, otherwise we're risking to miscalculate the total number at some point

Antreesy replied:
while ommiting these mutations here, PR might be reduced to:
- having a getter in conversationsStore (e.g. returning unreadMessagesMap, which includes:
`[token]: { unreadMessages: number, unreadMentions: boolean, unreadMentionDirect: boolean }`
  - with a logic similar to `getLastMessageMap()` function from composable (archived conversations are considered correctly, +1 in one conversation and -1 in another conversation are considered correctly);
  - less performant, but also less error-prone (and we're already iterating the list for similar purposes in `useDocumentTitle`, maybe result could also be reused in both places?)
- having an initialization logic (e.g. in App.vue or LeftSidebar.vue), which watch after this getter changes and emit events (with total payload) to Desktop client (or another recipients, where needed)

Antreesy commented on src/store/conversationsStore.js:
```diff
@@ -232,6 +264,14 @@ const mutations = {
+		this.dispatch('unreadCount/updateConversationCounts', {
+			token,
+			unreadMessages: state.conversations[token].unreadMessages || 0,
+			unreadMentions: state.conversations[token].unreadMention || 0,
+			unreadMentionDirect: state.conversations[token].unreadMentionDirect || 0,
```
these two are either `true` or `0`. I'd treat them as boolean only. I wouldn't want to rely on `true - 0 === 1` and `0 - true === -1`

Antreesy reviewed (commented):
Gave it a smoke test, seems to be working as expected 🎉

Leaving my thoughts for this PR. I suggest you to wait for another devs opinions, before rewriting anything; then we can find a common ground before putting more efforts into this feature.

markusstoll commented:
Here as with talk-desktop:
I do not want anyone to rush things, but I understood correctly that we shall wait for more input from your devs?

ShGKme commented:
Thank you for your contribution and moving forward to adding the counter-badge in the desktop client!

It seems to work, but I agree with @Antreesy's points.

We should not create a new store state for the unread counters, as this is not an independent state, but a computation on top of the existing state (conversation list). Having it as a new state will likely lead to inconsistent application state in the future and requires manual updates.

This makes sense for performance. However, even with ~1000 conversations counting, there should be no performance issue.

Let's simplify it to one getter in the existing `conversationsStore` and then watch for its changes to emit the global event in the `App.vue`.

ShGKme commented:
Hey, @markusstoll!

Do you plan to adjust the PRs? Maybe we can assist with anything?

This is an important feature that we'd like to add. If you don't have time now - no problem, but then we'd need to take it over.

markusstoll commented:
Sorry, I was drawn into some urgent issues in my paid work. I will start the rework soon, hopefully this week

markusstoll commented:
Hi,

I hope I understood your comments correctly. Pls review the reworked PR - will submit a matching PR for talk-desktop asap

ShGKme commented on src/App.vue:
```diff
@@ -95,6 +96,30 @@ export default {
 	},
 
 	computed: {
+		// Computed properties for unread counts
```
This comment doesn't have additional value. The code below is `computed`-s because it's literally the next line after `computed: {` and they are for `unread`-s as their names say.
```suggestion
```

ShGKme commented on src/App.vue:
```diff
@@ -95,6 +96,30 @@ export default {
 	},
 
 	computed: {
+		// Computed properties for unread counts
+		getTotalUnreadMessages() {
+			if (!this.$store.state.conversationsStore?.conversations) {
+				return 0;
+			}
+			return Object.values(this.$store.state.conversationsStore.conversations).reduce((total, conv) => {
+				return total + (conv.unreadMessages || 0);
+			}, 0);
+		},
+		
+		getTotalUnreadMentions() {
+			if (!this.$store.state.conversationsStore?.conversations) {
+				return 0;
+			}
+			return Object.values(this.$store.state.conversationsStore.conversations).filter(conv => conv.unreadMention).length;
+		},
+		
+		getTotalUnreadMentionsDirect() {
+			if (!this.$store.state.conversationsStore?.conversations) {
+				return 0;
+			}
+			return Object.values(this.$store.state.conversationsStore.conversations).filter(conv => conv.unreadMentionDirect).length;
+		},
+
```
Could you also use `conversation` instead of `conv`? Except for very common cases, we always use full names for variables.

ShGKme replied:
> However, getters in the store took into account archived conversations.
> 
> Here we also need to filter out archived conversations. You can do it by checking `!conversation.isArchived`.

@markusstoll This is still actual

ShGKme commented on src/App.vue:
```diff
@@ -619,6 +661,19 @@ export default {
+		 * Emits the UNREAD_COUNT_UPDATED event with the current counter values
+		 */
+		emitUnreadCountUpdated() {
+			const eventData = {
+				totalUnreadMessages: this.getTotalUnreadMessages,
+				totalUnreadMentions: this.getTotalUnreadMentions,
+				totalUnreadMentionsDirect: this.getTotalUnreadMentionsDirect
+			}
+
+			emit(EVENTS.UNREAD_COUNT_UPDATED, eventData)
```
To simplify naming a bit and remove the constant (see comment in the `talk-desktop`)
- The entire event payload is about counters, the total number of unread _something_, so we can omit it from every property name
- Let's add `conversations` in case a user would like to see the number of conversations, but not total messages in all the conversations
- Event is renamed to `unreads` to be plural (it's "unreads" that were updated)

```suggestion
			emit('talk:unreads:updated', {
				conversations: /* Number of conversations with unread messages */,
				messages: this.getTotalUnreadMessages,
				mentions: this.getTotalUnreadMentions,
				mentionsDirect: this.getTotalUnreadMentionsDirect
			})
```

ShGKme commented on src/App.vue:
```diff
@@ -163,6 +188,17 @@ export default {
 			}
 		},
+
+		// Watch for changes in unread counters and emit events
```
Same as with computed-s above
```suggestion
```

markusstoll commented:
Hi, I did take into account all your remarks and fixed them where needed.

On 
```
It didn't work without emitting here? On updates it's supposed to change conversations data and trigger computing
```
Indeed without initial counter is not shown on badge

On
```
May I ask if this was AI-generated? :)
```

You bet, as being a Java Developer for 25 years, I am mostly illiterate on such fancy JavaScript stuff and need AI support for all this syntax. But here I forgot to remove these additions in conversationsStore.js after deciding for smaller approach - I apologise for that.

After waiting for this feature for 1.5 years, I finally had to do something about it, using the means I have.

ShGKme commented:
> Indeed without initial counter is not shown on badge

In initial load (`mounted`) it could be necessary, but you also have it in `conversations-received` event handler.

ShGKme commented on src/App.vue:
```diff
@@ -619,6 +651,18 @@ export default {
 			}
 		},
+
+		/**
+		 * Emits the UNREAD_COUNT_UPDATED event with the current counter values
+		 */
+		emitUnreadCountUpdated() {
+			emit('talk:unread:updated', {
```
As per the comment above
```suggestion
		emitUnreadCountUpdated() {
			emit('talk:unreads:updated', {
```

ShGKme reviewed (commented):
Some comments above are still unresolved

markusstoll commented:
you were right, but pls note the spelling difference, which I changed back to 

```
emit('talk:unreads:updated', {
```

so it matches the doc in constants and the usage in talk-desktop

archived conversations are now ignored

Antreesy reviewed (commented):
Nice enhancements! Tested with following suggestions, to reduce amount of computations.

Feel free to squash all commits into one, and name it according to convention spec. something like `feat: <what's it doing>` should be fine. Then I'll trigger the CI to get it merged

ShGKme commented on src/App.vue:
```diff
@@ -95,6 +95,32 @@ export default {
 	},
 
 	computed: {
+		getTotalUnreadMessages() {
+			return this.$store.getters.conversationsList
+				.filter(conversation => !conversation.isArchived)
+				.reduce((total, conversation) => {
+					return total + (conversation.unreadMessages || 0);
+				}, 0);
+		},
+		
+		getTotalUnreadMentions() {
+			return this.$store.getters.conversationsList
+				.filter(conversation => !conversation.isArchived && conversation.unreadMention)
+				.length;
+		},
+		
+		getTotalUnreadMentionsDirect() {
+			return this.$store.getters.conversationsList
+				.filter(conversation => !conversation.isArchived && conversation.unreadMentionDirect)
+				.length;
+		},
+
+		getUnreadConversationCount() {
+			return this.$store.getters.conversationsList
+				.filter(conversation => !conversation.isArchived && (conversation.unreadMessages || 0) > 0)
+				.length;
+		},
```
In addition, with 4 watches it emitted event 4 times with the same on update.

markusstoll commented:
Thanks for your comments which made this change even more lean and clean.
I followed your advice, but changed to my proposed event name "talk:unread:updated" which is used in my PR on talk-desktop. As the details in the event data bear the plural I think we it makes sense to stick with singular in the event name.

Antreesy reviewed (approved):
Thanks, Markus 🚀

Antreesy commented:
/backport to stable31
